File: annotated.py
# NotSoTinyWM is written by Aleksi Torhamo <aleksi@torhamo.net>.
# It is based on TinyWM, which is written by Nick Welch <nick@incise.org>.
#
# This software is in the public domain and is provided AS IS, with NO WARRANTY.

# This is a lightly (in comparison to annotated.c) commented version of nstwm.py.
# For more in-depth explanations, refer to annotated.c.
# Why make code changes once, when you could make them two.. I mean four times instead!

from Xlib.display import Display
from Xlib import X, XK
from util import *

def decorate(win):
    # Normally windows are destroyed if their parent is destroyed.
    # Since we reparent windows to out decorations, they would normally be destroyed when the window manager dies.
    # Adding windows to our save-set prevents them being destroyed when we die.
    win.change_save_set(X.SetModeInsert)
    # Remove any borders the window might have, since it's gonna have decorations now..
    # XXX: I don't know if the X11 borders are purely a visual thing, of if they can actually be used for something..
    win.configure(border_width = 0)
    # Get window geometry so that we know how big to make the decorations
    geom = win.get_geometry()
    # Create the window, 5px borders plus a 20px bar at the top.
    # Setting the background pixel means we don't have to do any drawing and stuff ourselves.
    # We set the event mask for the window so that we get clicks only from the frame,
    # clicks to the window still go to the application as they're supposed to.
    frame = root.create_window(geom.x - 5, geom.y - 25, geom.width + 10, geom.height + 30,
        0, scr.root_depth, X.CopyFromParent, scr.root_visual, background_pixel = scr.white_pixel, event_mask = X.ButtonPressMask|X.ButtonReleaseMask)
    # Make sure window is over the frame. XXX: Is this really needed?
    frame.configure(sibling = win, stack_mode = X.Above)
    # Reparent the window to our decorations, take borders and top bar into account for the position.
    win.reparent(frame, 5, 25)
    # Make our frame (and the window) visible.
    frame.map()

dpy = Display()
scr = dpy.screen()
root = scr.root

# Grab the alt+F1 key combination on the root window, which means it'll only be reported to us
root.grab_key(dpy.keysym_to_keycode(XK.string_to_keysym("F1")),
    X.Mod1Mask, 1, X.GrabModeAsync, X.GrabModeAsync)

# Add decorations to existing windows
for win in root.query_tree().children:
    attr = win.get_attributes()
    # Skip adding decorations to this window if it's override_redirect or unmapped.
    # XXX: Why these two conditions, I have no idea. Every window manager seems
    # to use something along these lines. Somebody please explain. :(
    # Apparently, override_redirect is meant for "temporary pop-up windows that
    # should not be reparented or affected by the window manager's layout policy",
    # whatever that means. Something tells me I have never seen such a window, ever..
    if attr.override_redirect or attr.map_state == X.IsUnmapped:
        continue
    # The window seems sane, add decorations to it
    decorate(win)

while 1:
    # Get the next event from the queue
    ev = dpy.next_event()

    # ev.window is the window we asked for events, and ev.child is the window that received the event.

    # In our case, if ev.child == X.NONE, the event happened on the root window,
    # since that's where we asked for events. We ignore those, since we're not
    # really interested on clicks on the root window. And raising the root window
    # wouldn't make sense either..

    # We have only asked key press events for one key, so we can be sure any
    # key press event is for alt+F1 without explicitly checking for it.
    if ev.type == X.KeyPress and ev.child != X.NONE:
        # Raise the window in question to the top
        ev.child.configure(stack_mode = X.Above)
    # We got a mouse button press, start dragging. We only get clicks from frames
    # of windows we have decorated.
    elif ev.type == X.ButtonPress:
        # Ask for motion events so we can track the drag
        ev.window.grab_pointer(1, X.PointerMotionMask,
            X.GrabModeAsync, X.GrabModeAsync, X.NONE, X.NONE, X.CurrentTime)
        # Save the original position and size of the window
        attr = ev.window.get_geometry()
        # Save the start position of the drag
        start = ev
    # We only get motion events after we've started dragging, so no special
    # checking necessary.
    elif ev.type == X.MotionNotify:
        # Compress motion notify events, since it doesn't make any sense to
        # look at anything besides the most recent one. This speeds up resizing
        # considerably.
        # TODO: I think in theory we should stop if we see any button events,
        # since doing it like this can process them out of order. I doubt it
        # makes any difference in practice, however.
        ev = compress_motion(dpy, ev)
        # See how the pointer has moved relative to the root window.
        xdiff = ev.root_x - start.root_x
        ydiff = ev.root_y - start.root_y
        # Move or resize the window depending on which button was pressed.
        if start.detail == 1:
            start.window.configure(
                x = attr.x + xdiff,
                y = attr.y + ydiff
            )
        elif start.detail == 3:
            start.window.configure(
                width = max(1, attr.width + xdiff),
                height = max(1, attr.height + ydiff)
            )
    # We got a mouse button release, stop dragging.
    elif ev.type == X.ButtonRelease:
        # Stop receiving motion events
        dpy.ungrab_pointer(X.CurrentTime)

File: annotated.c
/* NotSoTinyWM is written by Aleksi Torhamo <aleksi@torhamo.net>.
 * It is based on TinyWM, which is written by Nick Welch <nick@incise.org>.
 *
 * This software is in the public domain and is provided AS IS, with NO WARRANTY. */

/* this is a commented version of nstwm.c, since uncommented code is much
 * easier to edit (when you're familiar with the code).
 */

/* most X stuff will be included with Xlib.h, but a few things require other
 * headers, like Xmd.h, keysym.h, etc.
 */
#include <X11/Xlib.h>

#define MAX(a, b) ((a) > (b) ? (a) : (b))

int main(void)
{
    Display * dpy;
    XWindowAttributes attr;

    /* we use this to save the pointer's state at the beginning of the
     * move/resize.
     */
    XButtonEvent start;

    XEvent ev;
    int xdiff, ydiff;

    /* return failure status if we can't connect */
    if(!(dpy = XOpenDisplay(0x0))) {
        return 1;
    }

    /* we use DefaultRootWindow to get the root window, which is a somewhat
     * naive approach that will only work on the default screen.  most people
     * only have one screen, but not everyone.  if you run multi-head without
     * xinerama then you quite possibly have multiple screens. (i'm not sure
     * about vendor-specific implementations, like nvidia's)
     *
     * many, probably most window managers only handle one screen, so in
     * reality this isn't really *that* naive.
     *
     * if you wanted to get the root window of a specific screen you'd use
     * RootWindow(), but the user can also control which screen is our default:
     * if they set $DISPLAY to ":0.foo", then our default screen number is
     * whatever they specify "foo" as.
     */

    /* you could also include keysym.h and use the XK_F1 constant instead of
     * the call to XStringToKeysym, but this method is more "dynamic."  imagine
     * you have config files which specify key bindings.  instead of parsing
     * the key names and having a huge table or whatever to map strings to XK_*
     * constants, you can just take the user-specified string and hand it off
     * to XStringToKeysym.  XStringToKeysym will give you back the appropriate
     * keysym or tell you if it's an invalid key name.
     *
     * a keysym is basically a platform-independent numeric representation of a
     * key, like "F1", "a", "b", "L", "5", "Shift", etc.  a keycode is a
     * numeric representation of a key on the keyboard sent by the keyboard
     * driver (or something along those lines -- i'm no hardware/driver expert)
     * to X.  so we never want to hard-code keycodes, because they can and will
     * differ between systems.
     */
    XGrabKey(dpy, XKeysymToKeycode(dpy, XStringToKeysym("F1")), Mod1Mask,
        DefaultRootWindow(dpy), True, GrabModeAsync, GrabModeAsync);

    /* XGrabKey and XGrabButton are basically ways of saying "when this
     * combination of modifiers and key/button is pressed, send me the events."
     * so we can safely assume that we'll receive Alt+F1 events, Alt+Button1
     * events, and Alt+Button3 events, but no others.  You can either do
     * individual grabs like these for key/mouse combinations, or you can use
     * XSelectInput with KeyPressMask/ButtonPressMask/etc to catch all events
     * of those types and filter them as you receive them.
     */
    XGrabButton(dpy, 1, Mod1Mask, DefaultRootWindow(dpy), True,
        ButtonPressMask|ButtonReleaseMask|PointerMotionMask, GrabModeAsync, GrabModeAsync, None, None);
    XGrabButton(dpy, 3, Mod1Mask, DefaultRootWindow(dpy), True,
        ButtonPressMask|ButtonReleaseMask|PointerMotionMask, GrabModeAsync, GrabModeAsync, None, None);

    start.subwindow = None;
    for(;;) {
        /* this is the most basic way of looping through X events; you can be
         * more flexible by using XPending(), or ConnectionNumber() along with
         * select() (or poll() or whatever floats your boat).
         */
        XNextEvent(dpy, &ev);

        /* this is our keybinding for raising windows.  as i saw someone
         * mention on the ratpoison wiki, it is pretty stupid; however, i
         * wanted to fit some sort of keyboard binding in here somewhere, and
         * this was the best fit for it.
         *
         * i was a little confused about .window vs. .subwindow for a while,
         * but a little RTFMing took care of that.  our passive grabs above
         * grabbed on the root window, so since we're only interested in events
         * for its child windows, we look at .subwindow.  when subwindow ==
         * None, that means that the window the event happened in was the same
         * window that was grabbed on -- in this case, the root window.
         */
        if(ev.type == KeyPress && ev.xkey.subwindow != None) {
            XRaiseWindow(dpy, ev.xkey.subwindow);
        } else if(ev.type == ButtonPress && ev.xbutton.subwindow != None) {
            /* we "remember" the position of the pointer at the beginning of
             * our move/resize, and the size/position of the window.  that way,
             * when the pointer moves, we can compare it to our initial data
             * and move/resize accordingly.
             */
            XGetWindowAttributes(dpy, ev.xbutton.subwindow, &attr);
            start = ev.xbutton;
        /* we only get motion events when a button is being pressed,
         * but we still have to check that the drag started on a window */
        } else if(ev.type == MotionNotify && start.subwindow != None) {
            /* here we "compress" motion notify events.
             *
             * if there are 10 of them waiting, it makes no sense to look at
             * any of them but the most recent.  in some cases -- if the window
             * is really big or things are just acting slowly in general --
             * failing to do this can result in a lot of "drag lag," especially
             * if your wm does a lot of drawing and whatnot that causes it to
             * lag.
             *
             * for window managers with things like desktop switching, it can
             * also be useful to compress EnterNotify events, so that you don't
             * get "focus flicker" as windows shuffle around underneath the
             * pointer.
             *
             * TODO: i think in theory we should stop if we see any button
             * events, since currently we can process them out of order.
             * i doubt it makes any difference in practice, however.
             */
            while(XCheckTypedEvent(dpy, MotionNotify, &ev));

            /* now we use the stuff we saved at the beginning of the
             * move/resize and compare it to the pointer's current position to
             * determine what the window's new size or position should be.
             *
             * if the initial button press was button 1, then we're moving.
             * otherwise it was 3 and we're resizing.
             *
             * we also make sure not to go negative with the window's
             * dimensions, resulting in "wrapping" which will make our window
             * something ridiculous like 65000 pixels wide (often accompanied
             * by lots of swapping and slowdown).
             *
             * even worse is if we get "lucky" and hit a width or height of
             * exactly zero, triggering an X error.  so we specify a minimum
             * width/height of 1 pixel.
             */
            xdiff = ev.xbutton.x_root - start.x_root;
            ydiff = ev.xbutton.y_root - start.y_root;
            if(start.button == 1) {
                XMoveWindow(dpy, start.subwindow,
                    attr.x + xdiff,
                    attr.y + ydiff
                );
            } else if(start.button == 3) {
                XResizeWindow(dpy, start.subwindow,
                    MAX(1, attr.width + xdiff),
                    MAX(1, attr.height + ydiff)
                );
            }
        } else if(ev.type == ButtonRelease) {
            start.subwindow = None;
        }
    }
}

